File: i18n.go
// Package webflow provides a wizard-style UI framework using HTML/WebView.
// This file contains the i18n support for backend-based translations.
package webflow

import (
	"encoding/json"
	"fmt"
	"sort"
	"strings"
	"sync"
)

// Translation markers - control characters for translation keys
const (
	// TranslationPrefix marks a string as a translation key (SOH - Start of Heading)
	TranslationPrefix = "\x01"
	// ArgSeparator separates key from args in translation strings (STX - Start of Text)
	ArgSeparator = "\x02"
)

// libraryTranslations contains built-in translations for 12 languages.
// Loaded from embedded JSON at init time.
var libraryTranslations map[string]map[string]string

func init() {
	if err := json.Unmarshal(translationsJSON, &libraryTranslations); err != nil {
		panic("webflow: failed to parse embedded translations: " + err.Error())
	}
}

// Package-level language state for immediate translation in T() and TF().
// This is set before rendering a page and used by T()/TF() to translate immediately.
var (
	currentLanguage        = "en"
	currentAppTranslations map[string]map[string]string
	langMu                 sync.RWMutex
)

// SetLanguage sets the current language and app translations for T() and TF().
// This must be called before rendering a page to ensure translations are correct.
func SetLanguage(lang string, appTrans map[string]map[string]string) {
	langMu.Lock()
	currentLanguage = lang
	currentAppTranslations = appTrans
	langMu.Unlock()
}

// GetLanguage returns the current UI language code.
func GetLanguage() string {
	langMu.RLock()
	defer langMu.RUnlock()
	return currentLanguage
}

// T translates a string immediately using the current language.
// Call SetLanguage() before rendering to ensure correct translations.
//
// Example:
//
//	button := Button{Label: T("button.next")} // Will be translated to "Next", "Weiter", etc.
func T(key string) string {
	langMu.RLock()
	lang := currentLanguage
	appTrans := currentAppTranslations
	langMu.RUnlock()

	return lookupTranslation(key, lang, appTrans)
}

// TF translates a key with format arguments, substituting placeholders immediately.
// Call SetLanguage() before rendering to ensure correct translations.
//
// Example:
//
//	title := TF("welcome.title", "Unison Auditor")
//	// "Welcome to the {0} Setup Wizard" → "Welcome to the Unison Auditor Setup Wizard"
func TF(key string, args ...any) string {
	langMu.RLock()
	lang := currentLanguage
	appTrans := currentAppTranslations
	langMu.RUnlock()

	template := lookupTranslation(key, lang, appTrans)

	// Substitute {0}, {1}, etc. with args
	for i, arg := range args {
		placeholder := fmt.Sprintf("{%d}", i)
		template = strings.ReplaceAll(template, placeholder, fmt.Sprint(arg))
	}
	return template
}

// TranslateString translates a string that may contain translation markers.
// If the string starts with TranslationPrefix (\x01), it's parsed as a translation key
// with optional arguments. Otherwise, the string is returned as-is.
//
// The lookup order is:
//  1. appTranslations[lang][key] - app-specific translation for current language
//  2. appTranslations["en"][key] - app-specific English fallback
//  3. libraryTranslations[lang][key] - built-in translation for current language
//  4. libraryTranslations["en"][key] - built-in English fallback
//  5. key itself - if no translation found
func TranslateString(s string, lang string, appTrans map[string]map[string]string) string {
	if s == "" || !strings.HasPrefix(s, TranslationPrefix) {
		return s // Literal text, return as-is
	}

	// Strip prefix and parse key + optional args
	content := s[1:] // Remove \x01
	var key string
	var args []any

	sepIndex := strings.Index(content, ArgSeparator)
	if sepIndex >= 0 {
		key = content[:sepIndex]
		argsJSON := content[sepIndex+1:]
		if err := json.Unmarshal([]byte(argsJSON), &args); err != nil {
			// Ignore parse errors, just use empty args
			args = nil
		}
	} else {
		key = content
	}

	// Get translation template with fallback chain
	template := lookupTranslation(key, lang, appTrans)

	// Recursively translate any args that are themselves translation keys
	for i, arg := range args {
		if str, ok := arg.(string); ok && strings.HasPrefix(str, TranslationPrefix) {
			args[i] = TranslateString(str, lang, appTrans)
		}
	}

	// Substitute placeholders: {0}, {1}, etc.
	result := template
	for i, arg := range args {
		placeholder := "{" + string(rune('0'+i)) + "}"
		argStr := ""
		switch v := arg.(type) {
		case string:
			argStr = v
		default:
			// Convert non-string args to string via JSON
			if b, err := json.Marshal(v); err == nil {
				argStr = string(b)
			}
		}
		result = strings.ReplaceAll(result, placeholder, argStr)
	}

	return result
}

// lookupTranslation finds the translation for a key with fallback chain.
// Order: appTrans[lang] -> appTrans["en"] -> libraryTranslations[lang] -> libraryTranslations["en"] -> key
func lookupTranslation(key, lang string, appTrans map[string]map[string]string) string {
	// Try app translations for current language first
	if appTrans != nil {
		if langMap, ok := appTrans[lang]; ok {
			if val, ok := langMap[key]; ok {
				return val
			}
		}
		// Fallback to app's English translation
		if lang != "en" {
			if langMap, ok := appTrans["en"]; ok {
				if val, ok := langMap[key]; ok {
					return val
				}
			}
		}
	}

	// Try library translations for current language
	if langMap, ok := libraryTranslations[lang]; ok {
		if val, ok := langMap[key]; ok {
			return val
		}
	}

	// Fallback to library's English translation
	if lang != "en" {
		if langMap, ok := libraryTranslations["en"]; ok {
			if val, ok := langMap[key]; ok {
				return val
			}
		}
	}

	// Return key itself as last resort
	return key
}

// appTranslations stores application-specific translations set via WithAppTranslations.
// These are merged with library translations on the frontend.
var appTranslations map[string]map[string]string

// LanguageInfo holds the code and display name for a language.
type LanguageInfo struct {
	Code string // e.g., "en", "de", "zh-Hans"
	Name string // e.g., "English", "Deutsch", "简体中文"
}

// GetAvailableLanguages returns all available languages sorted with English first,
// then the rest alphabetically by display name.
func GetAvailableLanguages() []LanguageInfo {
	var langs []LanguageInfo
	for code, trans := range libraryTranslations {
		name := code
		if n, ok := trans["_name"]; ok {
			name = n
		}
		langs = append(langs, LanguageInfo{Code: code, Name: name})
	}

	// Sort: English first, then alphabetically by name
	sort.Slice(langs, func(i, j int) bool {
		// English always first
		if langs[i].Code == "en" {
			return true
		}
		if langs[j].Code == "en" {
			return false
		}
		// Rest sorted alphabetically by name
		return langs[i].Name < langs[j].Name
	})

	return langs
}

// WithAppTranslations sets application-specific translations that are merged
// with the library's built-in translations. App translations take precedence.
//
// Example:
//
//	flow, _ := webflow.New(
//	    webflow.WithAppTranslations(map[string]map[string]string{
//	        "en": {"welcome.title": "Welcome to My App", "app.custom": "Custom string"},
//	        "de": {"welcome.title": "Willkommen bei Meiner App", "app.custom": "Eigene Zeichenkette"},
//	    }),
//	)
func WithAppTranslations(translations map[string]map[string]string) Option {
	return func(c *Config) {
		c.AppTranslations = translations
		appTranslations = translations // Also store in global for template access
	}
}

File: assets/i18n.js
// i18n.js - Frontend translation system for webflow
// Translations are injected by the backend from assets/translations.json

(function() {
    'use strict';

    const TRANSLATION_PREFIX = '\x01';
    const ARG_SEPARATOR = '\x02';

    // libraryTranslations is injected by backend before this script loads
    // appTranslations is also injected by backend (may be empty object)

    // Merged translations (library + app)
    const translations = {};

    // Current language (persisted to localStorage)
    let currentLang = 'en';
    try {
        currentLang = localStorage.getItem('webflow_lang') || 'en';
    } catch (e) {
        // localStorage may not be available in some contexts (WebViews, private browsing)
    }

    // Initialize translations by merging library and app translations
    function initTranslations() {
        // Start with library translations (injected by backend)
        if (typeof libraryTranslations !== 'undefined' && libraryTranslations) {
            for (const lang in libraryTranslations) {
                translations[lang] = { ...libraryTranslations[lang] };
            }
        }
        // Merge app translations (if provided by backend)
        if (typeof appTranslations !== 'undefined' && appTranslations) {
            for (const lang in appTranslations) {
                if (translations[lang]) {
                    translations[lang] = { ...translations[lang], ...appTranslations[lang] };
                } else {
                    translations[lang] = { ...appTranslations[lang] };
                }
            }
        }
    }

    // Translate a single text string
    function translate(text) {
        if (!text || typeof text !== 'string') {
            return text;
        }

        if (!text.startsWith(TRANSLATION_PREFIX)) {
            return text; // Literal text - return as-is
        }

        // Strip prefix and parse key + optional args
        const content = text.slice(1); // Remove \x01
        let key, args = [];

        const sepIndex = content.indexOf(ARG_SEPARATOR);
        if (sepIndex >= 0) {
            key = content.slice(0, sepIndex);
            try {
                args = JSON.parse(content.slice(sepIndex + 1));
                // Recursively translate any args that are themselves translation keys
                args = args.map(function(arg) {
                    if (typeof arg === 'string' && arg.startsWith(TRANSLATION_PREFIX)) {
                        return translate(arg);
                    }
                    return arg;
                });
            } catch (e) {
                // Ignore parse errors
            }
        } else {
            key = content;
        }

        // Get template string with fallback chain: current lang -> English -> raw key
        let template = translations[currentLang]?.[key]
                    || translations['en']?.[key]
                    || key;

        // Substitute placeholders: {0}, {1}, etc.
        return template.replace(/\{(\d+)\}/g, function(match, index) {
            const idx = parseInt(index, 10);
            return (args[idx] !== undefined) ? args[idx] : match;
        });
    }

    // Parse a translation string into key and args
    function parseTranslationString(text) {
        if (!text || !text.startsWith(TRANSLATION_PREFIX)) {
            return null;
        }
        const content = text.slice(1);
        const sepIndex = content.indexOf(ARG_SEPARATOR);
        if (sepIndex >= 0) {
            return {
                key: content.slice(0, sepIndex),
                args: content.slice(sepIndex + 1)
            };
        }
        return { key: content, args: null };
    }

    // Translate using parsed key/args object
    function translateParsed(parsed) {
        if (!parsed) return '';

        let args = [];
        if (parsed.args) {
            try {
                args = JSON.parse(parsed.args);
                // Recursively translate any args that are themselves translation keys
                args = args.map(function(arg) {
                    if (typeof arg === 'string' && arg.startsWith(TRANSLATION_PREFIX)) {
                        return translate(arg);
                    }
                    return arg;
                });
            } catch (e) { /* ignore */ }
        }

        let template = translations[currentLang]?.[parsed.key]
                    || translations['en']?.[parsed.key]
                    || parsed.key;

        return template.replace(/\{(\d+)\}/g, function(match, index) {
            const idx = parseInt(index, 10);
            return (args[idx] !== undefined) ? args[idx] : match;
        });
    }

    // Translate all text nodes and attributes in the page
    function translatePage() {
        // Walk DOM and translate text nodes containing prefix
        const walker = document.createTreeWalker(document.body, NodeFilter.SHOW_TEXT);
        while (walker.nextNode()) {
            const node = walker.currentNode;
            if (node.textContent && node.textContent.includes(TRANSLATION_PREFIX)) {
                // Parse and store as JSON for reliable round-trip through HTML attributes
                const parsed = parseTranslationString(node.textContent);
                if (parsed) {
                    const parent = node.parentElement;
                    if (parent && !parent.hasAttribute('data-i18n')) {
                        parent.setAttribute('data-i18n', JSON.stringify(parsed));
                    }
                    node.textContent = translateParsed(parsed);
                }
            }
        }

        // Translate placeholder and title attributes
        document.querySelectorAll('[placeholder], [title]').forEach(function(el) {
            if (el.placeholder && el.placeholder.startsWith(TRANSLATION_PREFIX)) {
                const parsed = parseTranslationString(el.placeholder);
                if (parsed) {
                    el.setAttribute('data-i18n-placeholder', JSON.stringify(parsed));
                    el.placeholder = translateParsed(parsed);
                }
            }
            if (el.title && el.title.startsWith(TRANSLATION_PREFIX)) {
                const parsed = parseTranslationString(el.title);
                if (parsed) {
                    el.setAttribute('data-i18n-title', JSON.stringify(parsed));
                    el.title = translateParsed(parsed);
                }
            }
        });
    }

    // Re-translate page using stored keys (for language change)
    function retranslatePage() {
        // Re-translate elements with stored keys
        document.querySelectorAll('[data-i18n]').forEach(function(el) {
            try {
                const parsed = JSON.parse(el.getAttribute('data-i18n'));
                el.textContent = translateParsed(parsed);
            } catch (e) { /* ignore malformed */ }
        });

        // Re-translate attributes
        document.querySelectorAll('[data-i18n-placeholder]').forEach(function(el) {
            try {
                const parsed = JSON.parse(el.getAttribute('data-i18n-placeholder'));
                el.placeholder = translateParsed(parsed);
            } catch (e) { /* ignore */ }
        });
        document.querySelectorAll('[data-i18n-title]').forEach(function(el) {
            try {
                const parsed = JSON.parse(el.getAttribute('data-i18n-title'));
                el.title = translateParsed(parsed);
            } catch (e) { /* ignore */ }
        });
    }

    // Set current language
    function setLanguage(lang) {
        if (translations[lang]) {
            currentLang = lang;
            try {
                localStorage.setItem('webflow_lang', lang);
            } catch (e) {
                // localStorage may not be available in some contexts
            }
        }
    }

    // Get current language
    function getLanguage() {
        return currentLang;
    }

    // Get language name for display
    function getLanguageName(lang) {
        return translations[lang]?.['_name'] || lang;
    }

    // Get available languages
    function getAvailableLanguages() {
        return Object.keys(translations).map(function(lang) {
            return {
                code: lang,
                name: translations[lang]?.['_name'] || lang
            };
        });
    }

    // Expose functions globally
    window.i18n = {
        init: initTranslations,
        translate: translate,
        translatePage: translatePage,
        retranslatePage: retranslatePage,
        setLanguage: setLanguage,
        getLanguage: getLanguage,
        getLanguageName: getLanguageName,
        getAvailableLanguages: getAvailableLanguages,
        TRANSLATION_PREFIX: TRANSLATION_PREFIX,
        ARG_SEPARATOR: ARG_SEPARATOR
    };

    // Note: initTranslations() is called from runtime.js after libraryTranslations and appTranslations are defined
})();
